Updating setup section, adding a build of book on pull requests (#42)
pangeo-data/foss4g-2022: tutorial/before/setup.md

# Setup: how to run the tutorial

This tutorial goal is to provide a wide introduction to the Pangeo ecosystem. 
While participants will see all the core libraries and possibilities offered by Pangeo software ecosystem, it is also important they get some insights of what is a Pangeo platform and how to use it. 
This is why we put in place such a platform on top of EOSC federation thanks to CESNET provider. 

> The recommended way to run this tutorial is thus to use this infrastructure, which is mainly a Dask-gateway enabled Jupyterhub proposing some pre-cooked software environment. 

Reviewer: ```suggestion
We recommend then running this tutorial using the EOSC/CESNET infrastructure, which is mainly a Dask-gateway enabled Jupyterhub proposing some pre-cooked software environment. 
```

For more details on this deployment, you may want to take a look at [Daskhub helm chart](https://github.com/dask/helm-chart/tree/main/daskhub).

This pages will describe three ways of running the tutorial:

> - Through the provided Pangeo platform deployed on EOSC resources (recommended),

Reviewer: ```suggestion
- Through the provided Pangeo platform deployed on EOSC resources (recommended)
```

> - Locally on your personal computer,

Reviewer: ```suggestion
- Locally on your personal computer
```

> - Using various Binderhubs solutions.

Reviewer: ```suggestion
- Using various Binderhubs solutions
```

## Running the tutorial on the provided Pangeo platform

First, before the FOSS4G workshop, all the participants will have to create an EGI account and enroll to the Pangeo virtual organisation to have access to the platform.
There are only two simple steps to do:

> 1. Create an account on the development instance of EGI Check-In, clicking on: https://aai-dev.egi.eu/registry/co_petitions/start/coef:2. You can easily delegate authentication with academic or social account.

Reviewer: ```suggestion
1. Create an account on the development instance of EGI Check-In, clicking on: https://aai-dev.egi.eu/registry/co_petitions/start/coef:2. You can easily delegate authentication with academic or social account e.g. GitHub.
```

1. Enroll the "vo.pangeo.eu" Virtual Organisation by clicking on: https://aai-dev.egi.eu/registry/co_petitions/start/coef:290.

It's important to do these steps as early as possible in order for Virtual Organisation managers to approve your petition to join.

Once your enrollment to Pangeo VO is validated, just log out from EGI Check-in to refresh your account.

:::{note}
Before the start of the course, you need to register to the Pangeo EGI-ACE EOSC infrastructure.
:::

> Then, you just need to click on [this link](https://pangeo-xxlarge.vm.fedcloud.eu/jupyterhub/hub/user-redirect/git-pull?repo=https%3A//github.com/pangeo-data/foss4g-2022&urlpath=lab/tree/foss4g-2022/tutorial/pangeo101/&branch=main) to open a Jupyterlab on the correct folder for the tutorial.

Reviewer 2: We will probably use https://pangeo-foss4g.vm.fedcloud.eu/jupyterhub/ because we have grafana and everything in place to monitor das

Author: I've discussed by email with @j34ni about EOSC deployments, and this is not what I understood from our exchange. I understood we would probably used the xxlarge deployment because it was the more stable and it was working correctly.

As I've not followed everything, I may well be wrong about that!

Reviewer 2: It is not decided yet... ideally I would prefer pangeo-foss4g with additional nodes.

Author: I've replaced the link with the other one.

I've not had time to look closer into the infrastructure side, I'm still getting the inputs in order to be able to connect to our deployments admin side.

Under the hood, this will clone the https://github.com/pangeo-data/foss4g-2022 repository and put you in the correct folder for executing our tutorial.

From there, you can open and execute any notebooks used during the workshop, enjoy!

> Alternatively, you can also open notebooks one by one from the PANGEO 101 section of this Jupyterbook by clicking on the rocket icon on top of the notebook and choosing Jupyterhub option.

Reviewer: ```suggestion
Alternatively, you can also open notebooks one by one from the PANGEO 101 section of this Jupyterbook by clicking on the rocket icon 🚀 on top of the notebook and choosing Jupyterhub option.
```

:::{note}
After the FOSS4G workshop, you can continue to use the resources, but note that we would probably modify the service and things might not work from time to time. 

> We would appreciate you get in touch with us to give us some feedback and be informed of the platform evolution.

Reviewer: We might need to provide how to contact us, through issues? GH discussions?

Author: I'm open to suggestions here. Github seems a good choice for me.

Reviewer 2: I am not sure what is the most recommended way. Maybe via github issues to make it simple.

Author: I've opted for Github Discussions, as this looks more friendly for newcomers.

You can also contribtue to "the Pangeo show and tell" online meetups.
:::

## Running on your own computer

Most part of this tutorial has been designed to run on limited computer resources, so it's fine to run it on your laptop.
It's a bit more complicated as you'll have to install the software environment yourself. You also wont'be able to test real Cloud distributed processing with Dask gateway.

Steps to run this tutorial on your own computer are listed below and demonstrated through Linux commands:

1. git clone the foss4g-2022 repository.
```bash
git clone https://github.com/pangeo-data/foss4g-2022.git
```

> 2. Install the required software environment with Miniconda. This can take a few minutes.

Reviewer: ```suggestion
2. Install the required software environment with Miniconda. If you don't have conda, install it by following these instructions (see [here](https://docs.conda.io/en/latest/miniconda.html)). This can take a few minutes.
```

```bash
# If you don't already have Miniconda or Anaconda
curl -o miniconda.sh https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash miniconda.sh -f -b -p /opt/anaconda
/opt/anaconda/bin/conda clean -tipy

> rm -f miniconda.sh

Author: Yeah, I thought about that and precised above that this is just a demonstration on Linux, but that is probably not clear.

Anyway, we won't add the commands from the three major OS?

Should I make the point this is only an example with Linux clearer?

> rm -f miniconda.sh

Reviewer 2: Yes probably better to specify the steps described are for Linux.. I am not even sure it is necessary to detail how to install miniconda because I suspect Linux is not the most common among the participants.

> rm -f miniconda.sh

Author: I've emphasized the Linux command part, and remove the script to install Miniconda.

```bash
conda activate pangeo-foss4g
jupyter lab
```
4. Open a web browser and connect to the Jupyterlab provided URL (you should see it in the jupyter lab command outputs), something like: http://localhost:8888/lab?token=42fac6733c6854578b981bca3abf5152.
5. Navigate to foss4g-2022/tutorial/pangeo101/ using the file browser on the left side of the Jupyterlab screen.

## Running using a Binderhub deployment

[Binderhub](https://binderhub.readthedocs.io/en/latest/) is a cloud service that allows users to share reproducible interactive computing environments from code repositories. It is generally used to enable other users to easily run your own code through Jupyter notebooks. 
It is a really cool service offered for free by several organisations (MyBinder through Jupyter, GESIS, etc.).

It's probably the easiest way to execute notebooks in this repository, as you only have to do one click to arrive in a Jupyterlab with all the necessary libraries.
However, the hardware resources you'll be allowed to use will be quite small, so you won't be able to grasp the full potential of Pangeo software stack, and some parts of the notebooks will be unavailable.

> All the notebooks on the PANGEO 101 section, for example [](../pangeo101/xarray_introduction.ipynb) have a rocket icon at the top, from which you can select Binder button to just run this notebook on the [GESIS Binder service](https://notebooks.gesis.org/binder/).

Reviewer 2: ```suggestion
All the notebooks on the PANGEO 101 section, for example ["Handling multi-dimensional arrays with xarray"](../pangeo101/xarray_introduction.ipynb) have a rocket icon at the top, from which you can select Binder button to just run this notebook on the [GESIS Binder service](https://notebooks.gesis.org/binder/).
```
